{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SampleID</th>\n",
       "      <th>RunID</th>\n",
       "      <th>ReadLength</th>\n",
       "      <th>TotalBases</th>\n",
       "      <th>ProjectID</th>\n",
       "      <th>Instrument</th>\n",
       "      <th>LibraryID</th>\n",
       "      <th>LibraryLayout</th>\n",
       "      <th>LibrarySource</th>\n",
       "      <th>Platform</th>\n",
       "      <th>File</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Ecogroup</th>\n",
       "      <th>Unique_Organism</th>\n",
       "      <th>Rechecked_Ecogroup</th>\n",
       "      <th>Citation</th>\n",
       "      <th>Remapped_Ecogroups</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMEA1904331</td>\n",
       "      <td>ERR299206</td>\n",
       "      <td>200</td>\n",
       "      <td>5760780600</td>\n",
       "      <td>PRJEB1254</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>6811014</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>PRJEB1254/ERR299206.unmapped_marked_adapters.bam</td>\n",
       "      <td>Alticorpus geoffreyi</td>\n",
       "      <td>Deep Benthic</td>\n",
       "      <td>Alticorpus geoffreyi</td>\n",
       "      <td>Deep Benthic</td>\n",
       "      <td>Malinsky_2019_Phylogeny_Supplemental Paper</td>\n",
       "      <td>Deep Benthic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAMEA1904331</td>\n",
       "      <td>ERR299215</td>\n",
       "      <td>200</td>\n",
       "      <td>5754979800</td>\n",
       "      <td>PRJEB1254</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>6811014</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>PRJEB1254/ERR299215.unmapped_marked_adapters.bam</td>\n",
       "      <td>Alticorpus geoffreyi</td>\n",
       "      <td>Deep Benthic</td>\n",
       "      <td>Alticorpus macrocleithrum</td>\n",
       "      <td>Deep Benthic</td>\n",
       "      <td>Malinsky_2019_Phylogeny_Supplemental Paper</td>\n",
       "      <td>Deep Benthic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAMEA1904331</td>\n",
       "      <td>ERR295132</td>\n",
       "      <td>200</td>\n",
       "      <td>5607303800</td>\n",
       "      <td>PRJEB1254</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>6811014</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>PRJEB1254/ERR295132.unmapped_marked_adapters.bam</td>\n",
       "      <td>Alticorpus geoffreyi</td>\n",
       "      <td>Deep Benthic</td>\n",
       "      <td>Alticorpus peterdaviesi</td>\n",
       "      <td>Deep Benthic</td>\n",
       "      <td>https://www.cell.com/cms/10.1016/j.isci.2019.0...</td>\n",
       "      <td>Deep Benthic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SAMEA1877429</td>\n",
       "      <td>ERR266498</td>\n",
       "      <td>200</td>\n",
       "      <td>5152679800</td>\n",
       "      <td>PRJEB1254</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>6526386</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>PRJEB1254/ERR266498.unmapped_marked_adapters.bam</td>\n",
       "      <td>Alticorpus macrocleithrum</td>\n",
       "      <td>Deep Benthic</td>\n",
       "      <td>Andinoacara coeruleopunctatus</td>\n",
       "      <td>Riverine</td>\n",
       "      <td>skip</td>\n",
       "      <td>Deep Benthic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SAMEA1877429</td>\n",
       "      <td>ERR271651</td>\n",
       "      <td>200</td>\n",
       "      <td>5251306000</td>\n",
       "      <td>PRJEB1254</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>6526386</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>PRJEB1254/ERR271651.unmapped_marked_adapters.bam</td>\n",
       "      <td>Alticorpus macrocleithrum</td>\n",
       "      <td>Deep Benthic</td>\n",
       "      <td>Aristochromis christyi</td>\n",
       "      <td>Shallow Benthic</td>\n",
       "      <td>Ad Konings Book of Cichlids, pg 50; 2-10m dept...</td>\n",
       "      <td>Deep Benthic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>680</th>\n",
       "      <td>SAMEA4033302</td>\n",
       "      <td>ERR1757757</td>\n",
       "      <td>302</td>\n",
       "      <td>31711443560</td>\n",
       "      <td>PRJEB15289</td>\n",
       "      <td>HiSeq X Ten</td>\n",
       "      <td>DN461234K:D3</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>PRJEB15289/ERR1757757.unmapped_marked_adapters...</td>\n",
       "      <td>Tyrannochromis macrostoma</td>\n",
       "      <td>Shallow Benthic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Shallow Benthic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>681</th>\n",
       "      <td>SAMEA4032105</td>\n",
       "      <td>ERR1749547</td>\n",
       "      <td>302</td>\n",
       "      <td>33527928562</td>\n",
       "      <td>PRJEB15289</td>\n",
       "      <td>HiSeq X Ten</td>\n",
       "      <td>DN461234K:A3</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>PRJEB15289/ERR1749547.unmapped_marked_adapters...</td>\n",
       "      <td>Tyrannochromis nigriventer</td>\n",
       "      <td>Shallow Benthic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Shallow Benthic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>682</th>\n",
       "      <td>SAMEA4033301</td>\n",
       "      <td>ERR1749548</td>\n",
       "      <td>302</td>\n",
       "      <td>31361555118</td>\n",
       "      <td>PRJEB15289</td>\n",
       "      <td>HiSeq X Ten</td>\n",
       "      <td>DN461234K:B3</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>PRJEB15289/ERR1749548.unmapped_marked_adapters...</td>\n",
       "      <td>Tyrannochromis nigriventer</td>\n",
       "      <td>Shallow Benthic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Shallow Benthic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>683</th>\n",
       "      <td>SAMEA4032106</td>\n",
       "      <td>ERR1749549</td>\n",
       "      <td>302</td>\n",
       "      <td>34238240112</td>\n",
       "      <td>PRJEB15289</td>\n",
       "      <td>HiSeq X Ten</td>\n",
       "      <td>DN461234K:C3</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>PRJEB15289/ERR1749549.unmapped_marked_adapters...</td>\n",
       "      <td>Tyrannochromis nigriventer</td>\n",
       "      <td>Shallow Benthic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Shallow Benthic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>684</th>\n",
       "      <td>SAMEA3388855</td>\n",
       "      <td>ERR1081367</td>\n",
       "      <td>250</td>\n",
       "      <td>14060940500</td>\n",
       "      <td>PRJEB1254</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>14544796</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>PRJEB1254/ERR1081367.unmapped_marked_adapters.bam</td>\n",
       "      <td>Tyrannochromis nigriventer</td>\n",
       "      <td>Shallow Benthic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Shallow Benthic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>685 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         SampleID       RunID  ReadLength   TotalBases   ProjectID  \\\n",
       "0    SAMEA1904331   ERR299206         200   5760780600   PRJEB1254   \n",
       "1    SAMEA1904331   ERR299215         200   5754979800   PRJEB1254   \n",
       "2    SAMEA1904331   ERR295132         200   5607303800   PRJEB1254   \n",
       "3    SAMEA1877429   ERR266498         200   5152679800   PRJEB1254   \n",
       "4    SAMEA1877429   ERR271651         200   5251306000   PRJEB1254   \n",
       "..            ...         ...         ...          ...         ...   \n",
       "680  SAMEA4033302  ERR1757757         302  31711443560  PRJEB15289   \n",
       "681  SAMEA4032105  ERR1749547         302  33527928562  PRJEB15289   \n",
       "682  SAMEA4033301  ERR1749548         302  31361555118  PRJEB15289   \n",
       "683  SAMEA4032106  ERR1749549         302  34238240112  PRJEB15289   \n",
       "684  SAMEA3388855  ERR1081367         250  14060940500   PRJEB1254   \n",
       "\n",
       "              Instrument     LibraryID LibraryLayout LibrarySource  Platform  \\\n",
       "0    Illumina HiSeq 2000       6811014        PAIRED       GENOMIC  ILLUMINA   \n",
       "1    Illumina HiSeq 2000       6811014        PAIRED       GENOMIC  ILLUMINA   \n",
       "2    Illumina HiSeq 2000       6811014        PAIRED       GENOMIC  ILLUMINA   \n",
       "3    Illumina HiSeq 2000       6526386        PAIRED       GENOMIC  ILLUMINA   \n",
       "4    Illumina HiSeq 2000       6526386        PAIRED       GENOMIC  ILLUMINA   \n",
       "..                   ...           ...           ...           ...       ...   \n",
       "680          HiSeq X Ten  DN461234K:D3        PAIRED       GENOMIC  ILLUMINA   \n",
       "681          HiSeq X Ten  DN461234K:A3        PAIRED       GENOMIC  ILLUMINA   \n",
       "682          HiSeq X Ten  DN461234K:B3        PAIRED       GENOMIC  ILLUMINA   \n",
       "683          HiSeq X Ten  DN461234K:C3        PAIRED       GENOMIC  ILLUMINA   \n",
       "684  Illumina HiSeq 2000      14544796        PAIRED       GENOMIC  ILLUMINA   \n",
       "\n",
       "                                                  File  \\\n",
       "0     PRJEB1254/ERR299206.unmapped_marked_adapters.bam   \n",
       "1     PRJEB1254/ERR299215.unmapped_marked_adapters.bam   \n",
       "2     PRJEB1254/ERR295132.unmapped_marked_adapters.bam   \n",
       "3     PRJEB1254/ERR266498.unmapped_marked_adapters.bam   \n",
       "4     PRJEB1254/ERR271651.unmapped_marked_adapters.bam   \n",
       "..                                                 ...   \n",
       "680  PRJEB15289/ERR1757757.unmapped_marked_adapters...   \n",
       "681  PRJEB15289/ERR1749547.unmapped_marked_adapters...   \n",
       "682  PRJEB15289/ERR1749548.unmapped_marked_adapters...   \n",
       "683  PRJEB15289/ERR1749549.unmapped_marked_adapters...   \n",
       "684  PRJEB1254/ERR1081367.unmapped_marked_adapters.bam   \n",
       "\n",
       "                       Organism         Ecogroup  \\\n",
       "0          Alticorpus geoffreyi     Deep Benthic   \n",
       "1          Alticorpus geoffreyi     Deep Benthic   \n",
       "2          Alticorpus geoffreyi     Deep Benthic   \n",
       "3     Alticorpus macrocleithrum     Deep Benthic   \n",
       "4     Alticorpus macrocleithrum     Deep Benthic   \n",
       "..                          ...              ...   \n",
       "680   Tyrannochromis macrostoma  Shallow Benthic   \n",
       "681  Tyrannochromis nigriventer  Shallow Benthic   \n",
       "682  Tyrannochromis nigriventer  Shallow Benthic   \n",
       "683  Tyrannochromis nigriventer  Shallow Benthic   \n",
       "684  Tyrannochromis nigriventer  Shallow Benthic   \n",
       "\n",
       "                   Unique_Organism Rechecked_Ecogroup  \\\n",
       "0             Alticorpus geoffreyi       Deep Benthic   \n",
       "1        Alticorpus macrocleithrum       Deep Benthic   \n",
       "2          Alticorpus peterdaviesi       Deep Benthic   \n",
       "3    Andinoacara coeruleopunctatus           Riverine   \n",
       "4           Aristochromis christyi    Shallow Benthic   \n",
       "..                             ...                ...   \n",
       "680                            NaN                NaN   \n",
       "681                            NaN                NaN   \n",
       "682                            NaN                NaN   \n",
       "683                            NaN                NaN   \n",
       "684                            NaN                NaN   \n",
       "\n",
       "                                              Citation Remapped_Ecogroups  \n",
       "0           Malinsky_2019_Phylogeny_Supplemental Paper       Deep Benthic  \n",
       "1           Malinsky_2019_Phylogeny_Supplemental Paper       Deep Benthic  \n",
       "2    https://www.cell.com/cms/10.1016/j.isci.2019.0...       Deep Benthic  \n",
       "3                                                 skip       Deep Benthic  \n",
       "4    Ad Konings Book of Cichlids, pg 50; 2-10m dept...       Deep Benthic  \n",
       "..                                                 ...                ...  \n",
       "680                                                NaN    Shallow Benthic  \n",
       "681                                                NaN    Shallow Benthic  \n",
       "682                                                NaN    Shallow Benthic  \n",
       "683                                                NaN    Shallow Benthic  \n",
       "684                                                NaN    Shallow Benthic  \n",
       "\n",
       "[685 rows x 17 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read the excel file into a pandas dataframe\n",
    "df = pd.read_excel(\"/Users/kmnike/Desktop/VCF_Excel_Sheets/SampleDatabase_updated_ecogroups.xlsx\")\n",
    "\n",
    "# Create a dictionary to map Unique_Organism to Rechecked_Ecogroup\n",
    "eco_map = df.set_index('Unique_Organism')['Rassigned_Ecogroup'].to_dict()\n",
    "\n",
    "# Use map to apply the eco_map to Organism column and create a new column named \"Remapped_Ecogroups\"\n",
    "df['Remapped_Ecogroups'] = df['Organism'].map(eco_map)\n",
    "\n",
    "# Print the updated dataframe\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_excel(\"/Users/kmnike/Desktop/VCF_Excel_Sheets/test_out.xlsx\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    " df = pd.read_excel(self.sample_database, sheet_name = 'vcf_samples') # This line generates a pandas dataframe using the \"sample_database\" attribute so we can use it below:\n",
    "        # lot going on with the below lines.\n",
    "        # self.s_dt[self.s_dt.Ecogroup.isin(self.ecogroups)] is returning all rows where the self.ecogroups values are contained in the \"Ecogroup\" column of the excel sheet\n",
    "        # The .SampleID.unique() is filtering these rows to include only unique values in the SampleIDs column. However, this creates a numpy array so the pd.DataFrame wrapper around the whole thing converts this numpy array to a pandas dataframe. \n",
    "        # The to_csv(self.good_samples_txt, header = False, index = False) converts into a csv file that bcftools will be able to take as input. The index/header = False eliminate any index/header information, leaving only filtered samples. \n",
    "        # Note that the name of the output file is self.good_samples.txt which is where the file pathing for the output file is taken care of\n",
    "ecogroups = \n",
    "df_filtered = df[df.Ecogroup.isin(ecogroups)]\n",
    "        pd.DataFrame(self.df_filtered.SampleID.unique()).to_csv(self.good_samples_csv, header = False, index = False)\n",
    "        self.df_filtered['metadata_id'] = self.df_filtered['SampleID'] + \"_\" + self.df_filtered['Ecogroup']\n",
    "        self.df_filtered[['SampleID', 'metadata_id']].to_csv(self.metadata_csv, index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SampleID</th>\n",
       "      <th>Ecogroup_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMEA2661258</td>\n",
       "      <td>Shallow_Benthic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAMN06685759</td>\n",
       "      <td>Riverine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAMEA6147768</td>\n",
       "      <td>AC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SAMEA2661248</td>\n",
       "      <td>Deep_Benthic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SAMEA2661259</td>\n",
       "      <td>Shallow_Benthic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SAMEA3388870</td>\n",
       "      <td>Diplotaxodon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SAMEA4032099</td>\n",
       "      <td>Deep_Benthic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SAMEA4032103</td>\n",
       "      <td>Shallow_Benthic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SAMEA4033299</td>\n",
       "      <td>Shallow_Benthic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>SAMEA4032104</td>\n",
       "      <td>Shallow_Benthic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>SAMEA4033300</td>\n",
       "      <td>Shallow_Benthic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>SAMEA4032101</td>\n",
       "      <td>Shallow_Benthic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>SAMEA4033297</td>\n",
       "      <td>Shallow_Benthic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>SAMEA4032102</td>\n",
       "      <td>Shallow_Benthic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>SAMEA4033298</td>\n",
       "      <td>Shallow_Benthic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>SAMN06685712</td>\n",
       "      <td>Shallow_Benthic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>SAMN06685754</td>\n",
       "      <td>Utaka</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        SampleID       Ecogroup_2\n",
       "0   SAMEA2661258  Shallow_Benthic\n",
       "1   SAMN06685759         Riverine\n",
       "2   SAMEA6147768               AC\n",
       "3   SAMEA2661248     Deep_Benthic\n",
       "4   SAMEA2661259  Shallow_Benthic\n",
       "5   SAMEA3388870     Diplotaxodon\n",
       "6   SAMEA4032099     Deep_Benthic\n",
       "7   SAMEA4032103  Shallow_Benthic\n",
       "8   SAMEA4033299  Shallow_Benthic\n",
       "9   SAMEA4032104  Shallow_Benthic\n",
       "10  SAMEA4033300  Shallow_Benthic\n",
       "11  SAMEA4032101  Shallow_Benthic\n",
       "12  SAMEA4033297  Shallow_Benthic\n",
       "13  SAMEA4032102  Shallow_Benthic\n",
       "14  SAMEA4033298  Shallow_Benthic\n",
       "15  SAMN06685712  Shallow_Benthic\n",
       "16  SAMN06685754            Utaka"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load the two sheets into dataframes\n",
    "df1 = pd.read_excel('/Users/kmnike/Desktop/VCF_Excel_Sheets/SampleDatabase.xlsx', sheet_name='vcf_samples')\n",
    "df2 = pd.read_excel('/Users/kmnike/Desktop/VCF_Excel_Sheets/SampleDatabase_old.xlsx', sheet_name='vcf_samples')\n",
    "\n",
    "# Merge the two dataframes based on the SampleID column\n",
    "merged = pd.merge(df1, df2, on='SampleID', suffixes=('_1', '_2'))\n",
    "# merged\n",
    "\n",
    "# Find the rows where the Ecogroup values are different between the two sheets\n",
    "diff = merged[merged['Ecogroup_1'] != merged['Ecogroup_2']]\n",
    "\n",
    "# Print the rows where the Ecogroup values are different\n",
    "diff = diff[['SampleID', 'Ecogroup_2']].reset_index(drop=True)\n",
    "diff\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0      True\n",
      "1      True\n",
      "2      True\n",
      "3      True\n",
      "4      True\n",
      "5      True\n",
      "6      True\n",
      "7      True\n",
      "8      True\n",
      "9      True\n",
      "10     True\n",
      "11     True\n",
      "12     True\n",
      "13     True\n",
      "14     True\n",
      "15     True\n",
      "16     True\n",
      "17    False\n",
      "Name: SampleID, dtype: bool\n",
      "0     SAMEA2661258\n",
      "1     SAMN06685759\n",
      "2     SAMEA6147768\n",
      "3     SAMEA2661248\n",
      "4     SAMEA2661259\n",
      "5     SAMEA3388870\n",
      "6     SAMEA4032099\n",
      "7     SAMEA4032103\n",
      "8     SAMEA4033299\n",
      "9     SAMEA4032104\n",
      "10    SAMEA4033300\n",
      "11    SAMEA4032101\n",
      "12    SAMEA4033297\n",
      "13    SAMEA4032102\n",
      "14    SAMEA4033298\n",
      "15    SAMN06685712\n",
      "16    SAMN06685754\n",
      "17    SAMN06685755\n",
      "Name: SampleID, dtype: object\n"
     ]
    }
   ],
   "source": [
    "mismatch = pd.read_excel('/Users/kmnike/Desktop/VCF_Excel_Sheets/SampleDatabase.xlsx', sheet_name='Mismatch')\n",
    "matched_samples = mismatch['SampleID'].isin(diff['SampleID'])\n",
    "print(matched_samples)\n",
    "print(mismatch['SampleID'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "old = pd.read_csv('/Users/kmnike/Desktop/EigenFiles/csv_files/filtered_samples.csv')\n",
    "new = pd.read_csv('/Users/kmnike/Desktop/EigenFiles/csv_files/metadata_for_R.csv')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0       True\n",
       "1       True\n",
       "2      False\n",
       "3       True\n",
       "4       True\n",
       "       ...  \n",
       "315     True\n",
       "316     True\n",
       "317     True\n",
       "318     True\n",
       "319     True\n",
       "Name: metadata_id, Length: 320, dtype: bool"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "match = new['metadata_id'].isin(old['metadata_id'])\n",
    "matches"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SampleID</th>\n",
       "      <th>metadata_id_x</th>\n",
       "      <th>metadata_id_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMEA1904331</td>\n",
       "      <td>SAMEA1904331_Deep_Benthic</td>\n",
       "      <td>SAMEA1904331_Deep_Benthic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAMEA1877429</td>\n",
       "      <td>SAMEA1877429_Deep_Benthic</td>\n",
       "      <td>SAMEA1877429_Deep_Benthic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAMEA2661258</td>\n",
       "      <td>SAMEA2661258_Deep_Benthic</td>\n",
       "      <td>SAMEA2661258_Shallow_Benthic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SAMEA4032098</td>\n",
       "      <td>SAMEA4032098_Shallow_Benthic</td>\n",
       "      <td>SAMEA4032098_Shallow_Benthic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SAMN06685761</td>\n",
       "      <td>SAMN06685761_Shallow_Benthic</td>\n",
       "      <td>SAMN06685761_Shallow_Benthic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>315</th>\n",
       "      <td>SAMEA4033302</td>\n",
       "      <td>SAMEA4033302_Shallow_Benthic</td>\n",
       "      <td>SAMEA4033302_Shallow_Benthic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>316</th>\n",
       "      <td>SAMEA4032105</td>\n",
       "      <td>SAMEA4032105_Shallow_Benthic</td>\n",
       "      <td>SAMEA4032105_Shallow_Benthic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>317</th>\n",
       "      <td>SAMEA4033301</td>\n",
       "      <td>SAMEA4033301_Shallow_Benthic</td>\n",
       "      <td>SAMEA4033301_Shallow_Benthic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>318</th>\n",
       "      <td>SAMEA4032106</td>\n",
       "      <td>SAMEA4032106_Shallow_Benthic</td>\n",
       "      <td>SAMEA4032106_Shallow_Benthic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>319</th>\n",
       "      <td>SAMEA3388855</td>\n",
       "      <td>SAMEA3388855_Shallow_Benthic</td>\n",
       "      <td>SAMEA3388855_Shallow_Benthic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>320 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         SampleID                 metadata_id_x                 metadata_id_y\n",
       "0    SAMEA1904331     SAMEA1904331_Deep_Benthic     SAMEA1904331_Deep_Benthic\n",
       "1    SAMEA1877429     SAMEA1877429_Deep_Benthic     SAMEA1877429_Deep_Benthic\n",
       "2    SAMEA2661258     SAMEA2661258_Deep_Benthic  SAMEA2661258_Shallow_Benthic\n",
       "3    SAMEA4032098  SAMEA4032098_Shallow_Benthic  SAMEA4032098_Shallow_Benthic\n",
       "4    SAMN06685761  SAMN06685761_Shallow_Benthic  SAMN06685761_Shallow_Benthic\n",
       "..            ...                           ...                           ...\n",
       "315  SAMEA4033302  SAMEA4033302_Shallow_Benthic  SAMEA4033302_Shallow_Benthic\n",
       "316  SAMEA4032105  SAMEA4032105_Shallow_Benthic  SAMEA4032105_Shallow_Benthic\n",
       "317  SAMEA4033301  SAMEA4033301_Shallow_Benthic  SAMEA4033301_Shallow_Benthic\n",
       "318  SAMEA4032106  SAMEA4032106_Shallow_Benthic  SAMEA4032106_Shallow_Benthic\n",
       "319  SAMEA3388855  SAMEA3388855_Shallow_Benthic  SAMEA3388855_Shallow_Benthic\n",
       "\n",
       "[320 rows x 3 columns]"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "compare = pd.merge(new, old, on='SampleID')\n",
    "compare"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "compare['metadata_id_x'].equals(compare['metadata_id_y'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mcgrath",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
